pkg/payload/precondition/clusterversion/recommendedupdate.go

```diff
@@ -0,0 +1,120 @@
+package clusterversion
+
+import (
+	"context"
+	"fmt"
+
+	configv1 "github.com/openshift/api/config/v1"
+	configv1listers "github.com/openshift/client-go/config/listers/config/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/openshift/cluster-version-operator/lib/resourcemerge"
+	precondition "github.com/openshift/cluster-version-operator/pkg/payload/precondition"
+)
+
+// RecommendedUpdate checks if clusterversion is upgradeable currently.
+type RecommendedUpdate struct {
+	key    string
+	lister configv1listers.ClusterVersionLister
+}
+
+// NewRecommendedUpdate returns a new RecommendedUpdate precondition check.
+func NewRecommendedUpdate(lister configv1listers.ClusterVersionLister) *RecommendedUpdate {
+	return &RecommendedUpdate{
+		key:    "version",
+		lister: lister,
+	}
+}
+
+// Run runs the RecommendedUpdate precondition.
+// Returns PreconditionError when possible, if the requested target release is Recommended=False.
+func (ru *RecommendedUpdate) Run(ctx context.Context, releaseContext precondition.ReleaseContext) error {
+	clusterVersion, err := ru.lister.Get(ru.key)
+	if apierrors.IsNotFound(err) || meta.IsNoMatchError(err) {
+		return nil
+	}
+	if err != nil {
+		return &precondition.Error{
+			Nested:  err,
+			Reason:  "UnknownError",
+			Message: err.Error(),
+			Name:    ru.Name(),
+		}
+	}
+	for _, recommended := range clusterVersion.Status.AvailableUpdates {
+		if recommended.Version == releaseContext.DesiredVersion {
+			return nil
+		}
+	}
+
+	for _, conditionalUpdate := range clusterVersion.Status.ConditionalUpdates {
+		if conditionalUpdate.Release.Version == releaseContext.DesiredVersion {
+			for _, condition := range conditionalUpdate.Conditions {
+				if condition.Type == "Recommended" {
+					switch condition.Status {
+					case metav1.ConditionTrue:
+						return nil
+					case metav1.ConditionFalse:
+						return &precondition.Error{
+							Reason: condition.Reason,
+							Message: fmt.Sprintf("Update from %s to %s is not recommended:\n\n%s",
+								clusterVersion.Status.Desired.Version, releaseContext.DesiredVersion, condition.Message),
```

Comment:
Hmm, is `clusterVersion.Status.Desired.Version` already updated to point to our target?  I thought that only happened after we got through the precondition checks?